allauth/socialaccount/providers/microsoft/views.py

```diff
@@ -84,6 +95,34 @@ def complete_login(self, request, app, token, **kwargs):
             )
         )
         extra_data = _check_errors(response)
+
+        if self.fetch_photo:
+
+            photo_metadata_response = (
+                get_adapter()
+                .get_requests_session()
+                .get(
+                    self.photo_metadata_url,
+                    headers=headers,
+                )
+            )
+            if not photo_metadata_response.ok:
+                raise OAuth2Error("Request to user photo metadata failed")
+
+            photo_data_response = (
+                get_adapter()
+                .get_requests_session()
+                .get(
+                    self.photo_data_url,
+                    headers=headers,
+                )
+            )
+            if not photo_data_response.ok:
+                raise OAuth2Error("Request to user photo data failed")
+
+            extra_data['photo_metadata'] = photo_metadata_response.json()
+            extra_data['photo'] = base64.b64encode(photo_data_response.content).decode('utf-8')
```

Comment:
**Question**: I am not sure storing photo's as data URIs is in the `SocialAccount` table is a wise plan. Not sure about the Microsoft API, but photos can be really large, and turning that into data URIs results in an even bigger footprint.

Also see: https://stackoverflow.com/questions/695151/data-protocol-url-size-limitations

Also, always fetching the photo at login seems inefficient. I am really wondering if this should be builtin behavior, or whether this is best left as a project specific feature.

Reply:
> I am not sure storing photo's as data URIs is in the SocialAccount table is a wise plan ... uri size

There is a way to limit the uri size. 

Provided the user has any photo set up, Microsoft auto generates the 48x48 photo automatically and when making the api call you can specify 48x48 as a fixed size for example https://graph.microsoft.com/v1.0/me/photos/48x48/$value. A simple solution would be to just use that as the endpoint for the photo request and the equivelant for the metadata request. I have a feeling that endpoint is specifically designed to help get around this exact issue. 

> Always fetching the photo at login seems inefficient

This is a little trickier. I think if we're down to 48x48 we're back into reasonable data size so not sure if this is still a huge issue. I think we'd be talking 10 kilobytes at most. The downside of not requesting at login is that if the user photo changes then we would be using an old photo. There's no way to check if the user data has changed since last logon either as far as I can tell. I did check for a 'last updated' date in metadata and there didn't seem to be one. 

> I am really wondering if this should be builtin behavior, or whether this is best left as a project specific feature.

Of course that's really up to you. In favour of including this feature:

- We already provide get_avatar_url and for a project supporting multiple social logins this simply enables/makes the feature work equivelantly across all users and all platforms. 
- The solution works as users would expect it to and fits within the existing solution with minimal change

In favour of not including it:

- It feels a bit like a workaround and storing base64 photo data in the user object isn't quite the intent of the solution you've designed

Reply:
The methods exposed by `ProviderAccount` including `get_avatar_url` are intended to extract common information from the metadata payload of the account. In this case, the photo is clearly not available in the metadata, and I would prefer not to forcefully store it there. Even if we shrink it to 48x48 it all becomes quite project specific.

If you rely on this functionality within your project you can easily place that logic inside the `pre_social_login()` method of the social account adapter.  So, I prefer not to include this out of the box.

Reply:
Am I welcome to add an example of how to do so into docs?

Reply:
Use `pre_social_login()` et al easily becomes more than a simple few-liners example, it does not really fit well into the Microsoft provider documentation. Perhaps https://github.com/pennersr/django-allauth/discussions/categories/show-and-tell is a more suitable place?

Reply:
Ok if there's a space for me to put it in show and tell and we can include a one liner/link into the provider documentation just so it's able to be found if anyone needs it, then that sounds like a sensible plan.